### PR TITLE
Clear mouse capture when control removed.

### LINKED ISSUE #1203: PointerReleased event hits button after it has been detached from VisualTree
```
Exception object: 00007fa1a403dd08
Exception type:   System.InvalidOperationException
Message:          Control is not attached to visual tree.
InnerException:   <none>
StackTrace (generated):
00007FFD4174CC40 00007FA1C756DE0B Avalonia.Visuals.dll!Avalonia.VisualExtensions.GetRootAndPosition(Avalonia.VisualTree.IVisual)+0x1bb
    00007FFD4174CCC0 00007FA1C75754B9 Avalonia.Visuals.dll!Avalonia.VisualExtensions.PointToClient(Avalonia.VisualTree.IVisual, Avalonia.Point)+0x19
    00007FFD4174CCF0 00007FA1C7575397 Avalonia.Input.dll!Avalonia.Input.MouseDevice.GetPosition(Avalonia.VisualTree.IVisual)+0x187
    00007FFD4174CE00 00007FA1C7575194 Avalonia.Controls.dll!Avalonia.Controls.Button.OnPointerReleased(Avalonia.Input.PointerReleasedEventArgs)+0xe4
    00007FFD4174AC00 00007FA1C1CBDDAF System.Private.CoreLib.dll!System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()+0x1f
    00007FFD4174AC10 00007FA1C74E786D Avalonia.Interactivity.dll!Avalonia.Interactivity.RoutedEvent+<>c__DisplayClass25_0.<AddClassHandler>b__0(System.Tuple`2<System.Object,Avalonia.Interactivity.RoutedEventArgs>)+0xfd
    00007FFD4174D470 00007FA1C74E7637 Avalonia.Interactivity.dll!Avalonia.Interactivity.Interactive.RaiseEventImpl(Avalonia.Interactivity.RoutedEventArgs)+0x87
    00007FFD4174D4D0 00007FA1C74F5EE8 Avalonia.Interactivity.dll!Avalonia.Interactivity.Interactive.BubbleEvent(Avalonia.Interactivity.RoutedEventArgs)+0xd8
    00007FFD4174D510 00007FA1C74E7568 Avalonia.Interactivity.dll!Avalonia.Interactivity.Interactive.RaiseEvent(Avalonia.Interactivity.RoutedEventArgs)+0xa8
    00007FFD4174D530 00007FA1C7574EC7 Avalonia.Input.dll!Avalonia.Input.MouseDevice.MouseUp(Avalonia.Input.IMouseDevice, Avalonia.Input.IInputRoot, Avalonia.Point, Avalonia.Input.MouseButton, Avalonia.Input.InputModifiers)+0x177
    00007FFD4174D5A0 00007FA1C756DA82 Avalonia.Input.dll!Avalonia.Input.MouseDevice.ProcessRawEvent(Avalonia.Input.Raw.RawMouseEventArgs)+0x202
    00007FFD4174D650 00007FA1C7561378 Avalonia.Input.dll!Avalonia.Input.InputManager.ProcessInput(Avalonia.Input.Raw.RawInputEventArgs)+0x38
    00007FFD4174D670 00007FA1C7560E1B Avalonia.Gtk3.dll!Avalonia.Gtk3.WindowBaseImpl+<>c__DisplayClass84_0.<OnInput>b__0()+0x7b
    00007FFD4174D6B0 00007FA1C754ADBF Avalonia.Base.dll!Avalonia.Threading.JobRunner.RunJobs(System.Nullable`1<Avalonia.Threading.DispatcherPriority>)+0x6f
    00007FFD4174D700 00007FA1C754ACBC Avalonia.Gtk3.dll!Avalonia.Gtk3.Gtk3Platform+<>c__DisplayClass28_0.<Signal>b__0()+0x10c
    00007FFD4174D780 00007FA1C754AB48 Avalonia.Gtk3.dll!Avalonia.Gtk3.Interop.GlibTimeout.Handler(IntPtr)+0x88
    00007FFD4174DA90 00007FA1C754A9EF Avalonia.Gtk3.dll!Avalonia.Gtk3.Gtk3Platform.RunLoop(System.Threading.CancellationToken)+0x5f
    00007FFD4174DAD0 00007FA1C754A95B Avalonia.Base.dll!Avalonia.Threading.Dispatcher.MainLoop(System.Threading.CancellationToken)+0xbb
    00007FFD4174DB20 00007FA1C2363332 AvalonStudio.dll!AvalonStudio.App.Main(System.String[])+0x2d2
```
What happened there is a PointerReleased event that has reached a Button that was already removed from the visual tree
It seems that MouseDevice.MouseUp has received that button from GetInputElementsAt

See  @kekekeks for more info, can be reproduced on AvalonStudio

### PR DESCRIPTION
**Pull request by grokys**
When a captured control is removed from the visual tree, clear the mouse capture.

Note that our capture logic needs to be improved - it shouldn't be possible to capture a non-visible or non-enabled control. This doesn't address that but it does fix the crash in #1203.

Also, if a control has been detached from the visual tree but the scene hasn't yet been updated, don't return the control as part of a hit-test.

(Doesn't) Fix #1203

### CODE REVIEW

**Review by kekekeks — APPROVED** [review]
LGTM

**Comment by danwalmsley** [comment]
Doesn't fix #1203  but still a required fix, LGTM.

**Review by danwalmsley — APPROVED** [review]
LGTM

**Review by kekekeks — CHANGES_REQUESTED** [review]
It seems that DetachedFromVisualTree is not being called in time or not being called at all.

**Review by danwalmsley — CHANGES_REQUESTED** [review]
I added some debugging code to see if Capture = null happens, as @kekekeks says, its not getting called.

**Comment by kekekeks** [comment]
I see several places in AvalonStudio where `OnDetachedFromVisualTree` is overridden without calling the base class. I think we need to fire `DetachedFromVisualTree` event outside of `OnDetachedFromVisualTree`

**Comment by grokys** [comment]
I've moved raising the events to `OnAttachedToLogicalTreeCore`/`OnDetachedFromLogicalTreeCore` so they should always get raised even if the user overrides OnAttachedToLogicalTree` or `OnDetachedFromLogicalTree` and forgets to call the base implementation.

File: src/Avalonia.Input/MouseDevice.cs [hunk 1]
```diff
@@ -55,6 +72,7 @@ public Point Position
         /// </remarks>
         public virtual void Capture(IInputElement control)
         {
+            // TODO: Check visibility and enabled state before setting capture.
```
[the remark this answers is not in the document]
Comment (danwalmsley) [reply src/Avalonia.Input/MouseDevice.cs]: see https://github.com/AvaloniaUI/Avalonia/issues/1211

**Review by danwalmsley — APPROVED** [review]
LGTM and tested with AS

**Review by kekekeks — APPROVED** [review]

**Pull request merged**